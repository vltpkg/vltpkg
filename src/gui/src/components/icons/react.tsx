export const React = (props: React.SVGProps<SVGSVGElement>) => {
  return (
    <svg
      width="500"
      height="500"
      viewBox="0 0 500 500"
      fill="none"
      xmlns="http://www.w3.org/2000/svg"
      {...props}>
      <g clip-path="url(#clip0_2967_809)">
        <path
          d="m250.504 284.862c270.046 284.862 285.889 269.029 285.889 249.499c285.889 229.968 270.046 214.136 250.504 214.136c230.961 214.136 215.119 229.968 215.119 249.499c215.119 269.029 230.961 284.862 250.504 284.862z"
          fill="currentColor"
        />
        <path
          d="m250.501 321.96c355.363 321.96 440.371 289.522 440.371 249.509c440.371 209.495 355.363 177.058 250.501 177.058c145.639 177.058 60.6318 209.495 60.6318 249.509c60.6318 289.522 145.639 321.96 250.501 321.96z"
          stroke="currentColor"
          stroke-width="18.913"
        />
        <path
          d="m187.724 285.725c240.155 376.482 310.768 433.837 345.442 413.83c380.116 393.823 365.721 304.031 313.29 213.274c260.859 122.517 190.247 65.1621 155.572 85.1689c120.898 105.176 135.293 194.968 187.724 285.725z"
          stroke="currentColor"
          stroke-width="18.913"
        />
        <path
          d="m187.726 213.277c135.295 304.034 120.9 393.826 155.574 413.833c190.248 433.84 260.861 376.485 313.292 285.728c365.723 194.971 380.118 105.179 345.443 85.172c310.769 65.1652 240.157 122.52 187.726 213.277z"
          stroke="currentColor"
          stroke-width="18.913"
        />
      </g>
      <defs>
        <clipPath id="clip0_2967_809">
          <rect
            width="397"
            height="353"
            fill="currentColor"
            transform="translate(52 73)"
          />
        </clipPath>
      </defs>
    </svg>
  )
}
