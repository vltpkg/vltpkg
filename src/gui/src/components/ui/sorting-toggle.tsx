import { useState } from 'react'
import { motion } from 'framer-motion'
import {
  ArrowDownAz,
  ArrowDownZa,
  type LucideProps,
} from 'lucide-react'
import {
  Tooltip,
  TooltipContent,
  TooltipTrigger,
  TooltipProvider,
} from '@/components/ui/tooltip.jsx'

interface SortingToggleProps<T> {
  filteredItems: T[]
  setFilteredItems: (i: T[]) => void
  sortKey: keyof T
}

interface SortToggleButton {
  icon: (props: LucideProps) => React.ReactNode
  option: SortOption
  toolTipContent: string
}

type SortOption = 'ascending' | 'descending'

const SortingToggle = <T extends {}>({
  sortKey,
  filteredItems,
  setFilteredItems,
}: SortingToggleProps<T>) => {
  const [sortToggle, setSortToggle] =
    useState<SortOption>('ascending')

  const sortToggleButtons: SortToggleButton[] = [
    {
      icon: props => <ArrowDownAz {...props} />,
      option: 'ascending',
      toolTipContent: 'Sort ascending',
    },
    {
      icon: props => <ArrowDownZa {...props} />,
      option: 'descending',
      toolTipContent: 'Sort descending',
    },
  ]

  const handleSortToggleClick = (option: SortOption) => {
    const newSortToggle =
      sortToggle === option ?
        option === 'ascending' ?
          'descending'
        : 'ascending'
      : option

    setSortToggle(newSortToggle)

    if (newSortToggle === 'ascending') {
      sortAlphabeticallyAscending()
    } else {
      sortAlphabeticallyDescending()
    }
  }

  const sortAlphabeticallyAscending = () => {
    const sorted = [...filteredItems].sort((a, b) =>
      String(a[sortKey]).localeCompare(String(b[sortKey])),
    )
    setFilteredItems(sorted)
  }

  const sortAlphabeticallyDescending = () => {
    const sorted = [...filteredItems].sort((a, b) =>
      String(b[sortKey]).localeCompare(String(a[sortKey])),
    )
    setFilteredItems(sorted)
  }

  return (
    <div className="flex mx-3">
      <div className="relative flex items-center p-1 h-[2.5rem] bg-white dark:bg-black w-full rounded-sm border border-[1px] border-muted-foreground/25">
        {/* the toggle slider */}
        <motion.div
          className="absolute h-[2rem] w-[2rem] rounded-sm bg-muted border border-muted-foreground/20"
          animate={{
            left: sortToggle === 'ascending' ? 4 : 36,
          }}
        />

        {sortToggleButtons.map((button, idx) => (
          <TooltipProvider key={idx}>
            <Tooltip>
              <TooltipTrigger asChild>
                <div
                  onClick={() => {
                    handleSortToggleClick(button.option)
                  }}
                  className="cursor-pointer h-[2rem] w-[2rem] rounded-sm flex items-center justify-center z-[1]">
                  <button.icon
                    size={20}
                    className="text-muted-foreground"
                  />
                </div>
              </TooltipTrigger>
              <TooltipContent>{button.toolTipContent}</TooltipContent>
            </Tooltip>
          </TooltipProvider>
        ))}
      </div>
    </div>
  )
}

export { SortingToggle }
