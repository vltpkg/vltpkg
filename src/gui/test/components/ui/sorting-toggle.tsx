import React from 'react'
import { vi, describe, it, expect, afterEach } from 'vitest'
import { cleanup, render } from '@testing-library/react'
import html from 'diffable-html'
import { SortingToggle } from '@/components/ui/sorting-toggle.jsx'

vi.mock('lucide-react', () => ({
  ArrowDownAz: 'gui-arrow-down-az-icon',
  ArrowDownZa: 'gui-arrow-down-za-icon',
}))

vi.mock('@/components/ui/tooltip.jsx', () => ({
  Tooltip: 'gui-tooltip',
  TooltipContent: 'gui-tooltip-content',
  TooltipTrigger: 'gui-tooltip-trigger',
  TooltipProvider: 'gui-tooltip-provider',
}))

expect.addSnapshotSerializer({
  serialize: v => html(v),
  test: () => true,
})

afterEach(() => {
  cleanup()
})

describe('sorting-toggle', () => {
  it('should render correctly', () => {
    const setStateMock = vi.fn()
    vi.spyOn(React, 'useState').mockReturnValue(['', setStateMock])

    const mockData = [
      {
        id: '01',
        name: 'mock-data-1',
      },
      {
        id: '02',
        name: 'mock-data-2',
      },
      {
        id: '03',
        name: 'mock-data-3',
      },
    ]

    const Container = () => {
      return (
        <SortingToggle
          filteredItems={mockData}
          setFilteredItems={setStateMock}
          sortKey="name"
        />
      )
    }

    const { container } = render(<Container />)

    expect(container.innerHTML).toMatchSnapshot()
  })
})
