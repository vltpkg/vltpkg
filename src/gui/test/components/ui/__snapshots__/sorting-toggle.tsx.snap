// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`sorting-toggle > should render correctly 1`] = `

<div class="flex mx-3">
  <div class="relative flex items-center p-1 h-[2.5rem] bg-white dark:bg-black w-full rounded-sm border border-[1px] border-muted-foreground/25">
    <div class="absolute h-[2rem] w-[2rem] rounded-sm bg-muted border border-muted-foreground/20">
    </div>
    <gui-tooltip-provider>
      <gui-tooltip>
        <gui-tooltip-trigger aschild="true">
          <div class="cursor-pointer h-[2rem] w-[2rem] rounded-sm flex items-center justify-center z-[1]">
            <gui-arrow-down-az-icon
              size="20"
              classname="text-muted-foreground"
            >
            </gui-arrow-down-az-icon>
          </div>
        </gui-tooltip-trigger>
        <gui-tooltip-content>
          Sort ascending
        </gui-tooltip-content>
      </gui-tooltip>
    </gui-tooltip-provider>
    <gui-tooltip-provider>
      <gui-tooltip>
        <gui-tooltip-trigger aschild="true">
          <div class="cursor-pointer h-[2rem] w-[2rem] rounded-sm flex items-center justify-center z-[1]">
            <gui-arrow-down-za-icon
              size="20"
              classname="text-muted-foreground"
            >
            </gui-arrow-down-za-icon>
          </div>
        </gui-tooltip-trigger>
        <gui-tooltip-content>
          Sort descending
        </gui-tooltip-content>
      </gui-tooltip>
    </gui-tooltip-provider>
  </div>
</div>

`;
